core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -338,12 +373,179 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal
     }
   }
 
+  /**
+   * Not thread safe. Assumed to only be called from the journal actor.
+   * However, unlike asyncWriteMessages it can be called before the previous Future completes
+   */
+  override def asyncDeleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
+
+    // TODO could "optimize" away deletes that overlap?
+    pendingDeletes.get(persistenceId) match {
```

Comment:
This is called from the outside, not only from receive, so `pendingDeletes` will have to be thread safe, no? Probably easiest to defer it to work in `self` somehow.

Reply:
Any of the methods that are implemented from the `AsyncWriteJournal` are called on the actor thread of the journal actor, we rely on this inside `asyncWriteMessages` where we use a regular Map to track concurrent writes.

I moved the delete code from a separate trait to stop it being used outside of a `CassandraJournal`, we had moved it out as we thought we were going to use it in `akka-update` but we took it out of scope in the end.

Reply:
Ah ok, I was under the misconception that you actually get the journal instance back from the persistence API, good then.

Reply:
That's the case for the read side. Write journals are purely internal to akka-persitsence